File: JavaScript/src/exercises.js
import crypto from "crypto";
export function change(amount) {
  if (amount < 0) {
    throw new RangeError("amount cannot be negative");
  }

  let quarters,
    dimes,
    nickels,
    pennies = 0;

  quarters = Math.floor(amount / 25);
  amount -= quarters * 25;
  dimes = Math.floor(amount / 10);
  amount -= dimes * 10;
  nickels = Math.floor(amount / 5);
  amount -= nickels * 5;
  pennies = amount / 1;
  amount -= pennies * 1;

  let change = [quarters, dimes, nickels, pennies];

  return change;
}

export function stretched(input) {
  const characters = Array.from(input);
  let stetchedWord = "";
  let spacesFound = 0;
  for (let i = 0; i < characters.length; i++) {
    if (characters[i] === " ") {
      spacesFound++;
    } else {
      stetchedWord += characters[i].repeat(i - spacesFound + 1);
    }
  }
  return stetchedWord;
}
export function say(word) {
  if (word == undefined) {
    return "";
  } else {
    return (next_word) => {
      if (next_word == undefined) {
        return word;
      } else {
        return say(word + " " + next_word);
      }
    };
  }
}
export function powers(base, limit, callback) {
  let i = 0;
  while (base ** i <= limit) {
    callback(base ** i);
    i++;
  }
}
export function* powersGenerator(base, limit) {
  let i = 0;
  while (base ** i <= limit) {
    yield base ** i;
    i++;
  }
  return;
}
export function makeCryptoFunctions({ forKey, using, withIV }) {
  let e = (encryptPhrase) => {
    let cipher = crypto.createCipheriv(using, forKey, withIV);
    let encrypt =
      cipher.update(encryptPhrase, "utf-8", "hex") + cipher.final("hex");
    return encrypt;
  };
  let d = (decryptPhrase) => {
    let decipher = crypto.createDecipheriv(using, forKey, withIV);
    let decrypted =
      decipher.update(decryptPhrase, "hex", "utf-8") + decipher.final("utf-8");
    return decrypted;
  };
  return [e, d];
}
export function topTenScorers(data) {
  data = Object.entries(data);
  data = data.flatMap((x) => x[1].map((p) => [...p, x[0]]));
  data = data.filter((x) => x[1] >= 15);
  data = data.map((p) => [
    {
      name: p[0],
      ppg: p[2] / p[1],
      team: p[3],
    },
  ]);
  data = data.flatMap((x) => x);
  // Used this website for sorting an array of objects in javascript
  // https://www.javascripttutorial.net/array/javascript-sort-an-array-of-objects/
  data = data.sort((a, b) => b.ppg - a.ppg);
  data = data.slice(0, 10);
  return data;
}
export async function pokemonInfo(pokeName) {
  // Used this website to learn how to use fetch with async and await
  // https://dmitripavlutin.com/javascript-fetch-async-await/
  const response = await fetch(
    `https://pokeapi.co/api/v2/pokemon/${pokeName}/`
  );
  const pokeInfo = await response.json();
  let data = {
    id: pokeInfo["id"],
    name: pokeInfo["name"],
    weight: pokeInfo["weight"],
  };
  return data;
}
export class Quaternion {
  #coef1;
  #coef2;
  #coef3;
  #coef4;
  constructor(num1, num2, num3, num4) {
    this.#coef1 = num1;
    this.#coef2 = num2;
    this.#coef3 = num3;
    this.#coef4 = num4;
  }
  coefficients() {
    return [this.#coef1, this.#coef2, this.#coef3, this.#coef4];
  }
  plus(otherQuar) {
    return new Quaternion(
      this.#coef1 + otherQuar.#coef1,
      this.#coef2 + otherQuar.#coef2,
      this.#coef3 + otherQuar.#coef3,
      this.#coef4 + otherQuar.#coef4
    );
  }
  times(otherQuar) {
    return new Quaternion(
      this.#coef1 * otherQuar.#coef1 -
        this.#coef2 * otherQuar.#coef2 -
        this.#coef3 * otherQuar.#coef3 -
        this.#coef4 * otherQuar.#coef4,
      this.#coef1 * otherQuar.#coef2 +
        this.#coef2 * otherQuar.#coef1 +
        this.#coef3 * otherQuar.#coef4 -
        this.#coef4 * otherQuar.#coef3,
      this.#coef1 * otherQuar.#coef3 -
        this.#coef2 * otherQuar.#coef4 +
        this.#coef3 * otherQuar.#coef1 +
        this.#coef4 * otherQuar.#coef2,
      this.#coef1 * otherQuar.#coef4 +
        this.#coef2 * otherQuar.#coef3 -
        this.#coef3 * otherQuar.#coef2 +
        this.#coef4 * otherQuar.#coef1
    );
  }
}

File: python/exercises.py
from cryptography.fernet import Fernet
import re
from dataclasses import dataclass
from typing import Tuple
import math

def change (amount: int) -> Tuple[int, int, int, int]:
    quarters = dimes = nickels = pennies = 0
    if not type (amount) is int:
        raise TypeError("Only integers are allowed")
    if amount < 0:
        raise ValueError("amount cannot be negative")

    quarters = math.floor (amount / 25)
    amount -= quarters * 25
    dimes = math.floor (amount / 10)
    amount -= dimes * 10
    nickels = math.floor (amount / 5)
    amount -= nickels * 5
    pennies = amount

    return (quarters, dimes, nickels, pennies)


def stretched(s: str, /) -> str:
    s: str = re.sub(r'\s+', '', s)
    # Got that line of code from https://bobbyhadz.com/blog/python-remove-spaces-tabs-newlines-from-string
    return_str: str = ""
    for index, letter in enumerate(s):
        return_str += (letter * (index + 1))
    return return_str


def powers(*, base: int, limit: int) -> list:
    i: int = 0
    while base ** i <= limit:
        yield base ** i
        i += 1
    return []


def say(word=None, /):
    if word is None:
        return ''
    return lambda next_word = None: word if next_word is None else say(word + " " + next_word)


def find_first_then_lower(predicate, words):
    if not words:
        raise ValueError("Empty list")
    for word in words:
        if predicate(word):
            return word.lower()
    raise ValueError("Nothing in list satisfies given property")


# Sorting function inspired by: https://www.programiz.com/python-programming/methods/list/sort
def comparator(list_item: float):
    list_item1: float = list_item[3]
    list_item2: float = list_item[2]
    return list_item1/list_item2


def top_ten_scorers(data: dict) -> list:
    result_list: list = []
    top_ten_list: list = []
    formatted_list: list = []

    for key in data:
        big_list: list = data[key]
        for small_list in big_list:
            small_list.insert(0, key)
            if small_list[2] >= 15:
                result_list.append(small_list)
                result_list.sort(reverse=True, key=comparator)
    top_ten_list = result_list[0:10]

    for small_list in top_ten_list:
        # To format to two decimal places: https://tutorial.eyehunts.com/python/how-to-display-2-decimal-places-in-python-example-code/#:~:text=Use%20str.,decimal%20places%20in%20the%20console.
        player_name: str = small_list[1]
        points_per_game: float = small_list[3]/small_list[2]
        team_name: str = small_list[0]
        formatted_list.append(
            f'{player_name}|{points_per_game:.2f}|{team_name}')

    return formatted_list


def crypto_functions():
    # info about fernet keys: https://www.geeksforgeeks.org/fernet-symmetric-encryption-using-cryptography-module-in-python/
    key = Fernet.generate_key()
    f = Fernet(key)

    def encode(data):
        return f.encrypt(data)

    def decode(token):
        return f.decrypt(token)

    return encode, decode


@dataclass(frozen=True)
class Quaternion:

    num1: int
    num2: int
    num3: int
    num4: int

    @property
    def coefficients(self):
        return (self.num1, self.num2, self.num3, self.num4)

    def __add__(self, otherQuaternion):
        return Quaternion(
            self.num1 + otherQuaternion.num1,
            self.num2 + otherQuaternion.num2,
            self.num3 + otherQuaternion.num3,
            self.num4 + otherQuaternion.num4)

    def __mul__(self, otherQuaternion):
        return Quaternion(
            self.num1 * otherQuaternion.num1 -
            self.num2 * otherQuaternion.num2 -
            self.num3 * otherQuaternion.num3 -
            self.num4 * otherQuaternion.num4,
            self.num1 * otherQuaternion.num2 +
            self.num2 * otherQuaternion.num1 +
            self.num3 * otherQuaternion.num4 -
            self.num4 * otherQuaternion.num3,
            self.num1 * otherQuaternion.num3 -
            self.num2 * otherQuaternion.num4 +
            self.num3 * otherQuaternion.num1 +
            self.num4 * otherQuaternion.num2,
            self.num1 * otherQuaternion.num4 +
            self.num2 * otherQuaternion.num3 -
            self.num3 * otherQuaternion.num2 +
            self.num4 * otherQuaternion.num1)
